File: app/src/main/java/com/android/sonu/ummsplash/data/DataManagerImpl.java
package com.android.sonu.ummsplash.data;

import android.content.Context;

import com.android.sonu.ummsplash.data.cache.NewPhotosCache;
import com.android.sonu.ummsplash.data.cache.PhotosCache;
import com.android.sonu.ummsplash.data.download.DownloadSession;
import com.android.sonu.ummsplash.di.ApplicationContext;
import com.android.sonu.ummsplash.ui.photo.Photo;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;

/**
 * Created by amanshuraikwar on 18/12/17.
 */

public class DataManagerImpl implements DataManager {

    @Inject
    @ApplicationContext
    Context applicationContext;

    @Inject
    DownloadSession downloadSession;

    @Inject
    NewPhotosCache newPhotosCache;

    @Inject
    public DataManagerImpl(){
    }


    @Override
    public Observable<List<Photo>> getMorePhotos(PhotosCache photosCache) {
        return photosCache.getMorePhotos();
    }

    @Override
    public Observable<List<Photo>> getCachedPhotos(PhotosCache photosCache) {
        return photosCache.getCachedPhotos();
    }

    @Override
    public boolean isCacheEmpty(PhotosCache photosCache) {
        return photosCache.isCacheEmpty();
    }

    @Override
    public NewPhotosCache getNewPhotosCache() {
        return newPhotosCache;
    }

    @Override
    public DownloadSession getDownloadSession() {
        return downloadSession;
    }
}

File: app/src/main/java/io/github/amanshuraikwar/splash/data/DataManagerImpl.kt
package io.github.amanshuraikwar.splash.data

import android.content.Context
import io.github.amanshuraikwar.splash.data.jetpack.JetpackDataManager
import io.github.amanshuraikwar.splash.data.jetpack.Listing
import io.github.amanshuraikwar.splash.data.network.NetworkDataManager
import io.github.amanshuraikwar.splash.di.ApplicationContext
import io.github.amanshuraikwar.splash.model.Photo
import io.github.amanshuraikwar.splash.ui.list.ListItem
import javax.inject.Inject

/**
 * Implementation of the Data Manager for the app.
 *
 * @author Amanshu Raikwar
 * Created by amanshuraikwar on 07/03/18.
 */
class DataManagerImpl @Inject constructor(
        @ApplicationContext val context: Context,
        private val networkDataManager: NetworkDataManager,
        private val jetpackDataManager: JetpackDataManager) : DataManager {

    override fun getAllPhotos(page: Int, orderBy: String, perPage: Int)
            = networkDataManager.getAllPhotos(page, orderBy, perPage)


    override fun getCurartedPhotos(page: Int, orderBy: String, perPage: Int)
            = networkDataManager.getCurartedPhotos(page, orderBy, perPage)


    override fun getCuratedPhotosListing(mapper: (Photo) -> ListItem<*>)
            = jetpackDataManager.getCuratedPhotosListing(mapper)
}